lib/util.js

```diff
@@ -885,7 +893,21 @@ function formatRaw(ctx, value, recurseTimes) {
   }
   ctx.seen.pop();
 
-  return reduceToSingleString(ctx, output, base, braces);
+  const res = reduceToSingleString(ctx, output, base, braces);
+  const budget = ctx.budget[ctx.indentationLvl] || 0;
+  const newLength = budget + res.length;
+  ctx.budget[ctx.indentationLvl] = newLength;
+  // If any indentationLvl exceeds this limit, limit further inspecting to the
+  // minimum. Otherwise the recursive algorithm might continue inspecting the
+  // object even though the maximum string size (~2 ** 28 on 32 bit systems and
+  // ~2 ** 30 on 64 bit systems) exceeded. The actual output is not limited at
+  // exactly 2 ** 27 but a bit higher. This depends on the object shape.
+  // This limit also makes sure that huge objects don't block the event loop
+  // significantly.
+  if (newLength > 2 ** 27) {
+    ctx.stop = true;
+  }
+  return res;
 }
```

Comment:
I get wanting to limit string construction to avoid memory issues and it looks like when this is implemented objects just won't be crawled _(something like `"[Object]"`)_ which are both good. 

* Am I correct that the indentation level is just a hint that there could be a possible problem _(it's unknown until the objects are crawled and inspected)_ and so it's being defensive and not going further?

* Can indentation level be translated to depth?
_(I think depth would be easier to reason about at a glance than indenting)_

* Can an object still be created that busts this heuristic?

Reply:
* Yes, it's unknown until we crawled and inspected the object to a point that the

Reply:
* Yes, it's unknown until we crawled and inspected the object and after exceeding the limit it's continuing defensively.
* Yes, that is possible. Using `depth` is not possible though, as it's value is set by the user, not by `inspect()` itself. `indentationLvl` is only meant to be internal.
* That would be possible. E.g., `util.inspect(new Uint8Array(1e8))`. If an individual part of an object exceeds the limit on it's own or the total space left, it's not going to protect against it.

Reply:
Looking at this again: the `depth` and the `indentationLvl` would be great to be combined but the `compact: true` mode has different indention for different values. That's why it's currently not possible :/

Reply:
That's cool. This is a good first attempt at least.